src/dict.js

```diff
@@ -157,47 +185,95 @@ Sk.builtin.dict.prototype.sq$contains = function (ob) {
 };
 
 Sk.builtin.dict.prototype.mp$ass_subscript = function (key, w) {
-    var k = kf(key);
-    var bucket = this.buckets[k.v];
-    var item;
+    const hash = kf(key);
+    let item;
+    if (typeof hash === "string") {
+        // we have a string so add to entries directly
+        item = this.entries[hash];
+        if (item === undefined) {
+            this.entries[hash] = { lhs: key, rhs: w };
+            this.size += 1;
+        } else {
+            item.rhs = w;
+        }
+    } else {
+        item = this.get$bucket_item(key, hash);
+        if (item === undefined) {
+            this.set$bucket_item(key, w, hash);
+            this.size += 1;
+        } else {
+            item.rhs = w;
+        }
+    }
+};
 
+Sk.builtin.dict.prototype.set$bucket_item = function (key, value, hash_value) {
+    // put a key, value pair into appropriate bucket and insert into entries
+    let key_hash,
+        bucket = this.buckets[hash_value];
+    const item = { lhs: key, rhs: value };
     if (bucket === undefined) {
-        // New bucket
-        bucket = {$hash: k, items: [
-            {lhs: key, rhs: w}
-        ]};
-        this.buckets[k.v] = bucket;
-        this.size += 1;
-        return;
+        bucket = this.buckets[hash_value] = [item];
+        key_hash = "#" + hash_value + "_" + 0; // this is the zeroth entry
+    } else {
+        // we might have a freeslot from deleting an item
+        // so either insert into freeslot or push
+        const free_slot_idx = bucket.indexOf(undefined);
+        if (free_slot_idx !== -1) {
+            key_hash = "#" + hash_value + "_" + free_slot_idx;
+            bucket[free_slot_idx] = item;
+        } else {
+            key_hash = "#" + hash_value + "_" + bucket.length;
+            bucket.push(item);
+        }
     }
+    this.entries[key_hash] = item;
+};
 
-    item = this.key$lookup(bucket, key);
-    if (item) {
-        item.rhs = w;
+
+
+
+Sk.builtin.dict.prototype.pop$bucket_item = function (key, hash_value) {
+    // pop a key, value pair and remove reference in buckets and entries
+    const bucket = this.buckets[hash_value];
+    let bucket_key, item;
+    if (bucket === undefined) {
         return;
     }
-
-    // Not found in dictionary
-    bucket.items.push({lhs: key, rhs: w});
-    this.size += 1;
+    for (let i = 0; i < bucket.length; i++) {
+        item = bucket[i];
+        if (item === undefined) {
+            continue;
+        }
+        bucket_key = item.lhs;
+        if (bucket_key === key || Sk.misceval.richCompareBool(key, bucket_key, "Eq")) {
+            const key_hash = "#" + hash_value + "_" + i;
+            delete this.entries[key_hash];
+            delete bucket[i]; // create empty slot for use in the event of reinsertion/collision
+            if (bucket.every((item) => item === undefined)) {
+                delete this.buckets[hash_value];
+            }
```

Comment:
here's the approach I took to check if a bucket was an `Array` of `empty` slots...
better approach?
`delete this.buckets[hash_value]` or just `bucket.length = 0`